api/src/main/java/org/apache/iceberg/util/WrapperSet.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.util;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterators;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.relocated.com.google.common.collect.Streams;
+
+/**
+ * A custom set for a {@link Wrapper} of the given type that maintains insertion order and does not
+ * allow null elements.
+ *
+ * @param <T> The type to wrap in a {@link Wrapper} instance.
+ */
+abstract class WrapperSet<T> implements Set<T> {
+  private final Set<Wrapper<T>> set = Sets.newLinkedHashSet();
+
+  protected WrapperSet(Iterable<Wrapper<T>> wrappers) {
+    wrappers.forEach(set::add);
+  }
+
+  protected WrapperSet() {}
+
+  protected abstract Wrapper<T> wrapper();
+
+  protected abstract Wrapper<T> wrap(T file);
+
+  protected abstract Class<T> elementClass();
+
+  protected interface Wrapper<T> {
+    T get();
+
+    Wrapper<T> set(T object);
+  }
+
+  protected Set<Wrapper<T>> set() {
+    return set;
+  }
+
+  @Override
+  public int size() {
+    return set.size();
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return set.isEmpty();
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public boolean contains(Object obj) {
```

Comment:
Given that `Set` allows `ClassCastException` in the API, I wonder whether we should simplify this.

```
@Override
public boolean contains(Object obj) {
  Preconditions.checkNotNull(obj, "Invalid object: null");
  Wrapper<T> wrapper = wrapper();
  boolean result = set.contains(wrapper.set(elementClass().cast(obj)));
  wrapper.set(null); // don't hold a reference to the value
  return result;
}
```

Up to you. Either way, we should use `elementClass().cast(obj)` to get rid of `@SuppressWarnings`.

Reply:
ah you're right and using `elementClass().cast(obj)` is indeed better here. If we remove the instance check here, then we should remove it everywhere where we use it